dbt_semantic_interfaces/parsing/where_filter/where_filter_dimension.py

```diff
@@ -0,0 +1,68 @@
+from __future__ import annotations
+
+from typing import List, Optional, Sequence
+
+from typing_extensions import override
+
+from dbt_semantic_interfaces.call_parameter_sets import TimeDimensionCallParameterSet
+from dbt_semantic_interfaces.parsing.where_filter.parameter_set_factory import (
+    ParameterSetFactory,
+)
+from dbt_semantic_interfaces.parsing.where_filter.query_interface import (
+    QueryInterfaceDimension,
+    QueryInterfaceDimensionFactory,
+)
+from dbt_semantic_interfaces.protocols.protocol_hint import ProtocolHint
+from dbt_semantic_interfaces.type_enums.time_granularity import TimeGranularity
+
+
+class WhereFilterDimension(ProtocolHint[QueryInterfaceDimension]):
+    """A dimension that is passed in through the where filter parameter."""
+
+    @override
+    def _implements_protocol(self) -> QueryInterfaceDimension:
+        return self
+
+    def __init__(  # noqa
+        self,
+        name: str,
+        entity_path: Sequence[str],
+        time_dimension_call_parameter_sets: List[TimeDimensionCallParameterSet],
+    ):
+        self.name = name
+        self.entity_path = entity_path
+        self._time_dimension_call_parameter_sets = time_dimension_call_parameter_sets
+        self.time_granularity: Optional[TimeGranularity] = None
+
+    def grain(self, time_granularity: str) -> QueryInterfaceDimension:
+        """The time granularity."""
+        self.time_granularity = TimeGranularity(time_granularity)
+        self._time_dimension_call_parameter_sets.append(
```

Comment:
Let me know if I'm getting out pocket on this. I'm likely biased from my personal preference for functional programming, but I found chain of passing the pointer to `time_dimension_factory.time_dimension_call_parameter_sets` odd. Additionally, until I followed the chain I was  confused as to why in the `where_filter_parser` dimension_parameter_sets were getting skipped unless they _didn't_ have a `time_granularity` specified. That is to say it appeared that the dimension_parameter_sets with `time_granularity` specified were getting trashed. I understand how it works now, and it's elegant in an OOP way. I do worry it's more complex than it needs to be though and will be harder for those unfamiliar with the code to approach. Additionally, thinking more about it, a question in my mind came up of what would happen if someone did the following
```python
{{ Dimension('<some_dimension>').grain('day').grain('week') }}
```
This chain wouldn't break on parsing and it would add the dimension to the `time_dimension_call_parameter_sets` twice, once with granularity `DAY` and once with granularity `WEEK`. Maybe that is what we want, but I don't think we do? My assumption here is that we'd actually just want it added once with granularity `WEEK`.

A way simplify things and to solve the duplication problem would be to not have `WhereFilterDimension`s directly modify the `time _dimension_call_parameter_sets`. Instead when `grain` is called it could just generate a `TimeDimensionCallParameterSet` and store it on a self property (maybe something like `self.time_dimension_call_parameter_set` which is defaulted to `None` on instantiation of the `WhereFilterDimsnion`. Then in ` parse_call_parameter_sets` when we iterate over `dimension_factory.created` we'd separate the created parameter sets into two lists, one of `DimensionCallParameterSet`s and the other of `TimeDimensionCallParameterSet`. The latter of which would be joined to the `time_dimension_factory.created` in the return statement.

Reply:
> I was confused as to why in the where_filter_parser dimension_parameter_sets were getting skipped unless they didn't have a time_granularity specified

I will add comments to this section and look for other opportunities to add comments.

> {{ Dimension('<some_dimension>').grain('day').grain('week') }}

Good call. I hadn't considered that edge case.

> A way simplify things and to solve the duplication problem...

Good idea. I had implemented something similar to this, but I wasn't sure if the order of `time_dimension_call_parameter_sets` mattered, so I refactored to this current approach. If we did what you are suggesting, then the `TimeDimensionCallParameterSet`s that are generated from this syntax `Dimension('metric_time').grain('day')` would always be appended after the ones that are generated from the `TimeDimension` syntax. Paul mentioned that the order doesn't matter (but I had already completed the refactor), so I think I will go ahead and implement what you have suggested.

Reply:
I believe I have addressed all of your concerns. Let me know if you would prefer a different approach.